### PR TITLE
[Kernel] Add GPU architecture guards to the CUTLASS w8a8 kernels to reduce binary size

### PR DESCRIPTION
**Pull request by tlrmchlsmth**
This PR wraps the CUTLASS kernel definitions in order to guard against compilation on architectures that will never use a particular kernel. The purpose of this is to reduce the size of the compiled binary. Each CUTLASS kernel is defined and optimized for specific GPU architectures, but each kernel is compiled for every arch defined in `CUDA_SUPPORTED_ARCHS`.

The normal way to deal with this is to look at the macro `__CUDA_ARCH__`. This macro is defined in the device-specific code but not on the host. All of our code runs on the host, so this PR uses these wrappers to "reach into" the device code to conditionally define the code.

### Results:
If I build 
```
export TORCH_CUDA_ARCH_LIST="7.0 7.5 8.0 8.6 8.9 9.0+PTX"
python setup.py build_ext --inplace
```
and then run the following (the strip is because I forgot to build in release mode):
```
strip vllm/_C.cpython-310-x86_64-linux-gnu.so
ls -al vllm/_C.cpython-310-x86_64-linux-gnu.so
```
main
```
-rwxrwxr-x 1 tms tms 146108048 Jun  4 20:49 vllm/_C.cpython-310-x86_64-linux-gnu.so
```
this PR:
```
-rwxrwxr-x 1 tms tms 142900880 Jun  4 20:04 vllm/_C.cpython-310-x86_64-linux-gnu.so
```

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by mgoin** [comment]
This seems reasonable to me (as in not too bad of a way to achieve this), is this in a land-able state given the passing checks?

**Comment by tlrmchlsmth** [comment]
> is this in a land-able state given the passing checks?

It's in a landable state. I want to measure the final size reduction for the wheel file when compiling for all CUDA ARCHS before marking it ready for review. I started to do this Friday but ran into some very long compile times

**Comment by tlrmchlsmth** [comment]
@mgoin @pcmoritz @comaniac @bnellnm ready for review!

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu [hunk 1]
```diff
@@ -48,9 +48,44 @@ using namespace cute;
...
+  template <typename... Args>
+  CUTLASS_DEVICE static void invoke(Args&&... args) {
+#if defined __CUDA_ARCH__ && __CUDA_ARCH__ >= 800 && __CUDA_ARCH__ < 890
+    Kernel::invoke(std::forward<Args>(args)...);
+#endif
+  }
+};
+
+template <typename Kernel>
+struct enable_sm89 : Kernel {
```
Comment (comaniac) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:77]:
```suggestion
struct enable_sm89_to_sm90 : Kernel {
```

**Review by comaniac — APPROVED** [review]

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**